feat: Add Ionos Cloud node driver support

dennislapchenko opened the pull request:
## Problem
<!-- Describe the root cause of the issue you are resolving. This may include what behavior is observed and why it is not desirable. If this is a new feature describe why we need this feature and how it will be used. -->

[Ionos Cloud](https://www.ionos.com/) is yet another cloud provider, it is lucrative performance and price-wise, therefore deploying RKE2 clusters with their written node driver would be really great.

 
## Solution
<!-- Describe what you changed to fix the issue. Relate your changes back to the original issue / feature and explain how this addresses the issue. -->
 - Added support for `ionoscloud_config` in `rancher2_machine_config_v2`
 - Added support for `ionoscloud_credential_config` in `rancher2_cloud_credential`
 - Added support for `add_cluster_credential`  in `rancher2_node_driver` (Ionoscloud's driver is adding its credential support upon installation) 
 - Updated credential config, machine config and node driver documentation
 
## Testing
<!-- Note: Confirm if the repro steps in the GitHub issue are valid, if not, please update the issue with accurate repro steps. -->

## Engineering Testing
### Manual Testing
<!-- Describe what manual testing you did (if no testing was done, explain why). -->
- Machine config was tested by creating configs and RKE2 clusters with various setups.
- 
### Automated Testing
<!--If you added/updated unit/integration/validation tests, describe what cases they cover and do not cover. -->
- Where examples are present - added the same type of tests (cloud credential)

## QA Testing Considerations
<!-- Highlight areas or (additional) cases that QA should test w.r.t a fresh install as well as the upgrade scenarios -->
 N/A
### Regressions Considerations
<!-- Dedicated section to specifically call out any areas that with higher chance of regressions caused by this change, include estimation of probability of regressions -->
N/A


## Notes

- Ionoscloud credential/machine configs accept either `username + password` or `token` fields, so technically one is optional if another is present (thus required). I am not sure how to correctly label what is `optional` and what is `required` in such case. Please advise
- Node drivers created via terraform have auto-generated ids. This is a definite issue when drivers activation is attempted during cloud credential and node pool creation. I have tried several approaches, with exclusions and/or node driver fetching by name if by id fails, but it all still blows somewhere downstream when cluster node pools are created. Therefore the only way to make it work is to create node driver by `kubectl apply`, which forces required `id` and then it correctly matches the defined `const ionoscloudConfigDriver` for all required credential/cluster operations. Would be nice to be able to set the `id` from terraform, but i do get why its not settable, as same nodedriver can be installed multiple times (for whatever reason). However it would still be nice to force the `id` from terraform when required. Any thoughts on this?


Kind regards

dennislapchenko committed:
feat: Add Ionos Cloud node driver support

dennislapchenko committed:
feat: Add Ionos Cloud node driver support

dennislapchenko committed:
feat: Reset default values for ionos machine fields

dennislapchenko committed:
feat: Reset default values for ionos machine fields

dennislapchenko commented:
@vbehar could anyone review this please?

matttrach commented:
Hello @dennislapchenko, thanks for contributing! sorry for the late response, is this still valuable?

matttrach commented:
You seem to have a conflict, please rebase on main and resolve.

dennislapchenko committed:
feat Add ionos-config to node templates

dennislapchenko committed:
feat Add ionos-config to node templates

dennislapchenko commented:
> Hello @dennislapchenko, thanks for contributing! sorry for the late response, is this still valuable?

Hey @matttrach thanks for getting back, better late then never! 😄 
Yes This is still valuable, using persona fork in production already, which is less than ideal.
Would be great to have this in. Rebased and fixed conflicts. Many thanks!

matttrach commented:
@dennislapchenko, this is failing to build. You can test building on your machine using the `make build` command.

matttrach commented:
@dennislapchenko, the build is failing, I can't merge as is. I have been merging in changes to master, but feel free to rebase. We need the build and tests to pass before I can get the team to review.

dennislapchenko committed:
feat Add ionos-config to node templates

dennislapchenko commented:
hey @matttrach thanks for merging in mains. Have committed fixes!

dennislapchenko committed:
feat Update RKE cloud init flag for latest version compatibility

matttrach commented:
Would you mind putting the config you tested with in the examples/resources/rancher2_machine_config_v2 directory?

dennislapchenko commented:
> Would you mind putting the config you tested with in the examples/resources/rancher2_machine_config_v2 directory?

hey @matttrach pasting the shortened config we use to setup this. let me know if you wanted the whole one to actually test or this is enough.

Also, how should i treat all the Copilot reviews, should that be addressed? It does seem that some of that is valid to check out. The somewhat of an issue with that is - you guys have changed the release process, so the old `release.yaml` file which i edited to publish to TFR is no longer doing anything. Feels like such a hassle to dissect new release actions to make it testable for me. Any advice on that? [here is how the edited release file looks](https://github.com/dennislapchenko/terraform-provider-rancher2/blob/8ce2f902362fb3fa9ab5edb3ed0ef2bef59a901b/.github/workflows/release.yaml)


```
resource "rancher2_machine_config_v2" "this" {
  for_each = {
    for config in var.machine_configs : config.name => config
  }

  generate_name = each.key

  ionoscloud_config {
    datacenter_id = var.datacenter_id
    lan_id        = coalesce(each.value.lan_id, var.lan_id)
    location      = coalesce(each.value.location, var.location)
    nat_id        = try(coalesce(each.value.nat_id, var.nat_id), null)

    server_type           = "CUBE"
    template              = each.value.server_type
    append_rke_cloud_init = true
    image                 = data.ionoscloud_image.this[each.key].id
    image_password        = var.console_user_passwd
    nic_dhcp              = !var.selfhosted_dhcp
    wait_for_ip_change    = false

    cloud_init_b64 = base64encode(templatefile(
      "${path.module}/cloudinit/cloudinit.yaml.tftpl",
      {
        console_user_name             = var.console_user_name
        console_user_passwd_encrypted = var.console_user_passwd_encrypted
        datacenter_id                 = var.datacenter_id
        ssh_port                      = var.ssh_port
        ssh_max_auth_tries            = var.ssh_max_auth_tries
        ldap_auth_dn                  = var.ldap_auth_dn
        ldap_auth_password            = var.ldap_auth_password
    }))
  }
}

locals {
  node_pools = [
    for pool in var.node_pools : merge(pool, {
       ...
      machine_config = {
        kind = rancher2_machine_config_v2.this[pool.machine_config_name].kind
        name = rancher2_machine_config_v2.this[pool.machine_config_name].name
      }
    })
  ]
}

module "rancher_cluster" {
  source     = "../../rancher/rancher-cluster"

  cluster_name       = var.cluster_name
  node_driver_cloud_credential_name = var.cloud_credential_id != "" ? var.cloud_credential_id : rancher2_cloud_credential.this[0].id
  node_pools                        = local.node_pools

}

matttrach commented:
1. Config:
The example config is just for reference at the moment, in the future it will be used to generate documentation. As our CI doesn't have access to IonosCloud it be unlikely to be actually run for testing anytime soon.
2. AI Code review:
The code review tool is meant to be helpful, not blocking. If it points out something which is valid and important, please follow its suggestions. If it suggests something outside of the scope of your change that doesn't effect the functionality of your code then feel free to respond saying so and resolve the conversation. Please minimally leave something saying why the suggestion isn't being implemented before resolving the conversation.
3. Publish to Terraform Registry
What are you publishing to the registry? You said "Feels like such a hassle to dissect new release actions to make it testable for me. ", are you releasing something to the registry in order to test?
Maybe check out what our test suite uses to test the checked out code rather than something published in the registry:
```
func WriteTerraformRc(t *testing.T, path string) error {
	repoRoot, err := filepath.Abs(g.GetRepoRootContext(t, t.Context(), ""))
	if err != nil {
		t.Fatalf("Error getting git root directory: %v", err)
	}
	formatString := `provider_installation {
  dev_overrides {
    "rancher/rancher2" = "%s/bin"
  }
  direct {
    exclude = []
  }
}`
	content := fmt.Sprintf(formatString, repoRoot)
	return os.WriteFile(path, []byte(content), 0644)
}
```
This essentially creates something like this `.terraform.rc` in the example directory where the test is run:
```
provider_installation {
  dev_overrides {
    "rancher/rancher2" = "/home/user/go/src/github.com/rancher/terraform-provider-rancher2/bin"
  }
  direct {
    exclude = []
  }
}
```
This tells Terraform to directly download all providers except the terraform-provider-rancher2 one, which it should get from the /home/user/go/src/github.com/rancher/terraform-provider-rancher2/bin directory. The bin directory can be automatically filled out using `make build`.  You can tell Terraform where your rc file is with the `TF_CLI_CONFIG_FILE` environment directory.
Hopefully this helps you test changes without needing to release your fork of the provider?

dennislapchenko commented:
@matttrach thanks, I do think the slop review has some valid points, need to spend time on checking it out.

regarding provider build - you are right that i should, as previously, test it by updating terraformrc file. The reason why i did publish to TFR is because we already use this functionality since the day this MR was created, and the process of having a private provider registry seemed too long so I just published to TFR.

dennislapchenko committed:
fix: address ionos review feedback and correct machine config field names

dennislapchenko commented:
@matttrach I have finally have found time to deal with this. copilots review had legit points, all were implemented.
I believe this is finally good to go! :)

dennislapchenko committed:
fix: harden CUBE sizing diff-suppress guard and drop stray comment

dennislapchenko committed:
fix: drop unused ionos const and dead disk_size CUBE branch

dennislapchenko commented on rancher2/structure_node_driver.go:
```diff
@@ -14,6 +14,7 @@ func flattenNodeDriver(d *schema.ResourceData, in *managementClient.NodeDriver)
 
 	d.SetId(in.ID)
 	d.Set("active", in.Active)
+	d.Set("add_cloud_credential", in.AddCloudCredential)
 	d.Set("builtin", in.Builtin)
 	d.Set("checksum", in.Checksum)
```
That split is correct and standard for this SDK: `d.Set` only returns an error on shape/type mismatches, which realistically only happen for lists/maps. For a scalar — and `add_cloud_credential` is a bool set from a bool — d.Set cannot fail. ADded line landed in exactly the right tier and matches all 8 sibling scalars.

dennislapchenko commented:
@matttrach hey! any info if this feature can be scoped into some upcoming release?
